### PR TITLE
Fix visibility of let expression's inner variables in the debugger view

### LINKED ISSUE #28883: [Debugger] `let` expression inner variables are visible outside the expression
**Description:**
Currently `let` expression variables are visible in debug mode. Those variables should not be visible.

**Steps to reproduce:**
- Consider the code segment:
```
public function main() {
    string greeting = let string hello = "Hello ",
                          string ballerina = "Ballerina!"
                      in hello + ballerina;
}
```
- Run the above program in debbubg mode.
- We can see the `greeting`, `hello` and `ballerina` are visible in the local variables panel.
- Expected behaviour is only `greeting` variable should be visible. Other variables (`hello` and `ballerina`) should not be visible

image

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #28883

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java [hunk 1]
```diff
@@ -1155,7 +1156,8 @@ public void visit(BLangFail failNode) {
     @Override
     public void visit(BLangSimpleVariableDef astVarDefStmt) {
         VarKind kind;
-        if (astVarDefStmt.var.symbol.origin == SymbolOrigin.VIRTUAL) {
+        if (astVarDefStmt.var.symbol.origin == SymbolOrigin.VIRTUAL ||
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:1159]: Shall we use separate VarKind for let expression? Should we sepearate out expression variable in bir @manuranga

Comment (manuranga) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:1159]: I don't this is right. `SYNTHETIC` kind is for variables that user didn't write themself. But if you look the original issue, both `hello` and `ballerina` are coming from the source, not desugered.  So why should they be marked `SYNTHETIC`.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java [hunk 2]
```diff
@@ -946,6 +946,17 @@ private void createLocalVariableTable(BIRFunction func, BIRVarToJVMIndexMap inde
                 if (localVar.endBB != null) {
                     endLabel = labelGen.getLabel(funcName + endBB.id.value + "beforeTerm");
                 }
+            } else if (localVar.kind == VarKind.SYNTHETIC && isCompilerAddedVars(localVar.metaVarName)) {
+                if (localVar.startBB != null) {
+                    startLabel = labelGen.getLabel(funcName + SCOPE_PREFIX + localVar.insScope.id);
+                }
+                for (int j = i + 1; j < func.localVars.size(); j++) {
```
Comment (manuranga) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:953]: does this var not have a `endBB`? If so, isn't that an issue.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java (hunk 1 above)
Comment (gabilang) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:1159]: We need to have a way to separately identify let expression inner variables among the function local variables. Currently let expression inner variables also marked with the variable kind `LOCAL` and we can't differentiate them from other local variables to separately handle their visibility scopes. Yes `SYNTHETIC` kind has been introduced for the variables generated at desugar (usually these variable names starts with specific symbols like $). So, using `SYNTHETIC` kind for  let expression inner variable might not be a good approach. Is it okay to add a new variable kind for expression inner variables?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java (hunk 2 above)
Comment (gabilang) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:953]: We need to end the visibility of inner variables once we came out the expression. `Let` expression does not have any terminator instructions and so we can't keep the end label based on the `endBB`. 
eg: For the following ballerina code, 
```
public function main() {

    int a = 1;

    string greeting = let string hello = "Hello ", string ballerina = "Ballerina!" 
                        in hello + ballerina;

    int b = 5;
}
```
BBs are generated like this,
```
bb0 {
        %1 = ConstLoad 1;
        %4 = ConstLoad Hello ;
        %6 = ConstLoad Ballerina!;
        %8 = %4 + %6;
        %3 = %8;
        %13 = ConstLoad 5;
        %0 = ConstLoad 0;
        GOTO bb2;
    }
    bb1 {
        GOTO bb2;
    }
    bb2 {
        return;
    }
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java [hunk 3]
```diff
@@ -946,6 +946,17 @@ private void createLocalVariableTable(BIRFunction func, BIRVarToJVMIndexMap inde
                 if (localVar.endBB != null) {
                     endLabel = labelGen.getLabel(funcName + endBB.id.value + "beforeTerm");
                 }
+            } else if (localVar.kind == VarKind.LET) {
+                if (localVar.startBB != null) {
+                    startLabel = labelGen.getLabel(funcName + SCOPE_PREFIX + localVar.insScope.id);
+                }
+                for (int j = i + 1; j < func.localVars.size(); j++) {
+                    BIRVariableDcl syntheticVar = func.localVars.get(j);
+                    if (syntheticVar.kind == VarKind.SYNTHETIC && !isCompilerAddedVars(syntheticVar.metaVarName)) {
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:955]: What happen if there are not variables after let expression variable. 
public function main() {
    string greeting = let string hello = "Hello ", string ballerina = "Ballerina!" 
                        in hello + ballerina;
}

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java (hunk 2 above)
Comment (manuranga) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/methodgen/MethodGen.java:953]: I believe the correct thing to do is to spit `bb0` to two bb's in the frontend (we also have to make sure our optimizer doesn't merge them)
eg:
```
    bb0 {
        %1 = ConstLoad 1;
        %4 = ConstLoad Hello ;
        %6 = ConstLoad Ballerina!;
        %8 = %4 + %6;
        %3 = %8;
    }
    bb1 {
        %13 = ConstLoad 5;
        %0 = ConstLoad 0;
        GOTO bb3;
    }
    bb2 {
        GOTO bb3;
    }
    bb3 {
        return;
    }
```

then we can use above `bb1` as the end bb.

**Comment by gabilang** [comment]
Closing this PR since the solution for this issue is needed to be discussed further.

**Pull request closed without merging**